src/operator/keycloak.ts

```diff
@@ -601,178 +552,149 @@ async function internalIdp(api: KeycloakApi, connection: KeycloakConnection) {
-        )
+        console.info(`Creating group ${groupName}`)
+        return api.groups.realmGroupsPost(keycloakRealm, group)
       }),
     )
     console.info('Finished managing groups')
-    await keycloakConfigMapChanges()
   } catch (error) {
     throw extractError('managing groups', error)
   }
 }
 
-export async function removeUserGroups(
-  api: { users: UsersApi; groups: GroupsApi },
-  existingUser: UserRepresentation,
-  teamGroups: string[],
-): Promise<void> {
-  try {
-    const { body: existingUserGroups } = await api.users.realmUsersIdGroupsGet(keycloakRealm, existingUser.id as string)
-
-    await Promise.all(
-      existingUserGroups.map(async (group) => {
-        if (!teamGroups.includes(group.name)) {
-          await api.users.realmUsersIdGroupsGroupIdDelete(keycloakRealm, existingUser.id as string, group.id as string)
-        }
-      }),
-    )
-  } catch (error) {
-    throw extractError('removing user groups', error)
-  }
-}
-
-export async function addUserGroups(
-  api: { users: UsersApi; groups: GroupsApi },
-  existingUser: UserRepresentation,
+export async function updateUserGroups(
+  api: KeycloakApi,
+  user: UserRepresentation,
+  groupsByName: Record<string, string>,
   teamGroups: string[],
 ): Promise<void> {
-  try {
-    const { body: currentKeycloakGroups } = await api.groups.realmGroupsGet(keycloakRealm)
-    const { body: existingUserGroups } = await api.users.realmUsersIdGroupsGet(keycloakRealm, existingUser.id as string)
-
-    await Promise.all(
-      teamGroups.map(async (teamGroup) => {
-        const existingGroup = existingUserGroups.find((el) => el.name === teamGroup)
-
-        if (!existingGroup) {
-          const groupId = currentKeycloakGroups.find((el) => el.name === teamGroup)?.id
-          if (groupId) {
-            await api.users.realmUsersIdGroupsGroupIdPut(keycloakRealm, existingUser.id as string, groupId as string)
-          }
+  const userId = user.id!
+  const existingUserGroups = (await api.users.realmUsersIdGroupsGet(keycloakRealm, userId)).body
+  const userGroupIds: Set<string> = new Set(existingUserGroups.map((userGroup) => groupsByName[userGroup.name]))
+  const teamGroupIds: Set<string> = new Set()
+  let groupUpdates = 0
+  await Promise.all(
+    teamGroups.map(async (teamGroup): Promise<void> => {
+      const teamGroupId = groupsByName[teamGroup]
+      if (teamGroupId) {
+        if (!userGroupIds.has(teamGroupId)) {
+          console.info(`Adding user ${user.email} to ${teamGroup}`)
+          await api.users.realmUsersIdGroupsGroupIdPut(keycloakRealm, userId, teamGroupId)
+          groupUpdates += 1
         }
-      }),
-    )
-  } catch (error) {
-    throw extractError('adding user groups', error)
+        teamGroupIds.add(teamGroupId)
+      } else {
+        console.info(`Group ${teamGroup} does not exist, skipping assignment`)
+      }
+    }),
+  )
+  await Promise.all(
+    existingUserGroups.map(async (userGroup): Promise<void> => {
```

Comment:
I would rather do a forEach loop here as we do not transform any data.
```suggestion
    existingUserGroups.forEach(async (userGroup): Promise<void> => {
```

Reply:
I would have preferred that too since we actually don't need the result. But unfortunately, `Array.forEach` returns really nothing, not even `Promise<void>`.

Reply:
Ah yes I see, thats sad :/